omniduct/filesystems/s3.py

```diff
@@ -0,0 +1,154 @@
+        # Check if still able to perform requests against AWS
+        import botocore
+        try:
+            self._client.list_buckets()
+        except botocore.exceptions.ClientError as e:
+            if len(e.args) > 0:
+                if 'ExpiredToken' in e.args[0] or 'InvalidToken' in e.args[0]:
+                    return False
+                elif 'AccessDenied' in e.args[0]:
+                    return True
+
+    def _disconnect(self):
+        pass
+
+    # File enumeration
+
+    def _exists(self, path):
+        if self.isfile(path):
+            return True
+        if self.isdir(path):
+            return True
+        return False
+
+    def _isdir(self, path):
+        response = next(iter(self.__dir_paginator(path)))
+        if 'CommonPrefixes' in response or 'Contents' in response:
+            return True
+        return False
+
+    def _isfile(self, path):
+        try:
+            self._client.get_object(Bucket=self.bucket, Key=path or '')
+            return True
+        except:
+            return False
+
+    def __dir_paginator(self, path):
+        path = path or ''
+        if path.endswith('/'):
+            path = path[:-1]
+        paginator = self._client.get_paginator('list_objects')
+        iterator = paginator.paginate(
+            Bucket=self.bucket,
+            Prefix=path + ('/' if path else ''),
+            Delimiter='/',
+            PaginationConfig={'PageSize': 500}
+        )
+        return iterator
+
+    def _listdir(self, path):
+        iterator = self.__dir_paginator(path)
+
+        paths = []
+        for response_data in iterator:
+            for prefix in response_data.get('CommonPrefixes', []):
+                paths.append(prefix['Prefix'][:-1])  # Remove trailing slash
+            for prefix in response_data.get('Contents', []):
+                paths.append(prefix['Key'])
+
+        return paths
+
+    def _showdir(self, path):
+        iterator = self.__dir_paginator(path)
+
+        paths = []
+        for response_data in iterator:
+            for prefix in response_data.get('CommonPrefixes', []):
+                paths.append({
+                    'type': 'dir',
+                    'name': prefix['Prefix'][:-1].split('/')[-1]  # Remove trailing slash
+                })
+            for prefix in response_data.get('Contents', []):
+                paths.append({
+                    'type': 'file',
+                    'name': prefix['Key'].split('/')[-1],
+                    'last_modified': prefix['LastModified'],
+                    'owner': prefix['Owner']['DisplayName'],
+                    'size': prefix['Size']
+                })
+
+        df = pd.DataFrame(paths)
+        col_order = [col for col in ('type', 'name', 'size', 'owner', 'last_modified') if col in df.columns]
+        return df[col_order]
+
+    def _find(self, pattern, path_prefix, files, dirs):
+        if dirs:
+            logger.warning('S3Client does not support finding directories.')
```

Comment:
add something like. "finding in all directories instead"

Reply:
Thanks. Fixed.